### PR TITLE
PXP-2570 Feat/data access privileges

### PR DESCRIPTION
**Pull request by vzpgb**
Please make sure to follow the [DEV guidelines](https://gen3.org/resources/developer/dev-introduction/) before asking for review.

### New Features
- Implemented auth command for getting user access privileges

### CODE REVIEW

File: gen3-client/g3cmd/auth.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+			endPointPostfix := "/user/user" // Information about current user
+
+			host, project_access, err := function.CheckPrivileges(profile, "", endPointPostfix, "application/json", nil)
+
+			if err != nil {
+				log.Fatalf("Fatal authentication error: %s\n", err)
+			} else {
+				if len(project_access) == 0 {
+					fmt.Printf("\nYou don't currently have access to data from any projects at %s\n", host)
```
Comment (mfshao) [inline gen3-client/g3cmd/auth.go]: if an ouput message is very meaningful you could use `log.Printf()` or `log.Println()`. In gen3-client I have set to let  these `log.x()` methods also write to a local log file in order to keep a record

File: gen3-client/g3cmd/auth.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+				} else {
+					fmt.Printf("\nYou have access to the following project(s) at %s:\n", host)
+
+					// Sort by project name
+					projects := make([]string, 0, len(project_access))
+					for project := range project_access {
+						projects = append(projects, project)
+					}
+					sort.Strings(projects)
```
Comment (mfshao) [inline gen3-client/g3cmd/auth.go]: Very nice here! Can we also sort for each premission in `project_access`?

File: gen3-client/jwt/functions.go [hunk 3]
```diff
@@ -184,8 +180,54 @@ func (f *Functions) DoRequestWithSignedHeader(profile string, configFileType str
...
+
+	_, resp, err := f.GetResponse(profile, configFileType, endpointPostPrefix, contentType, bodyBytes)
+	if err != nil {
+		return "", "", err
+	}
+
+	msg, err = f.ParseFenceURLResponse(resp)
+	return msg.Url, msg.GUID, err
+
+	panic("Unexpected case")
```
Comment (mfshao) [inline gen3-client/jwt/functions.go]: this code is unreachable

File: gen3-client/jwt/functions.go [hunk 4]
```diff
@@ -184,8 +180,54 @@ func (f *Functions) DoRequestWithSignedHeader(profile string, configFileType str
...
+	}
+
+	project_access, ok := data["project_access"].(map[string]interface{})
+	if !ok {
+		panic("Not possible to read user access privileges")
+	}
+
+	return host, project_access, err
+
 	panic("Unexpected case")
```
Comment (mfshao) [inline gen3-client/jwt/functions.go]: this code also becomes unreachable

File: gen3-client/jwt/functions.go [hunk 5]
```diff
@@ -164,16 +161,15 @@ func (f *Functions) DoRequestWithSignedHeader(profile string, configFileType str
 		if resp != nil && resp.StatusCode == 401 {
 			isExpiredToken = true
 		} else if err != nil {
```
Comment (mfshao) [inline gen3-client/jwt/functions.go]: probably can get rid of this `else if` ? Contents in `else if` and `else` look identical

File: gen3-client/jwt/functions.go [hunk 6]
```diff
@@ -184,8 +180,54 @@ func (f *Functions) DoRequestWithSignedHeader(profile string, configFileType str
...
+	host, resp, err := f.GetResponse(profile, configFileType, endpointPostPrefix, contentType, bodyBytes)
+	if err != nil {
+		return "", nil, err
+	}
+
+	str := ResponseToString(resp)
+
+	err = json.Unmarshal([]byte(str), &data)
+	if err != nil {
+		panic(err)
```
Comment (mfshao) [inline gen3-client/jwt/functions.go]: In general we should discourage to use `panic` unless there is a `defer-panic-recover` chain. Instead you can just use `log.Fatal` or `log.Print`.

I know there are still some `panic` leftovers in the current code. But they are going to be replaced eventually (and hopefully) so let's not introduce new members to them 😄 

Also, for this `err`, you could just return it and let the caller of this function to deal with it, just like all other errs in this function, to keep a consistent behavior

File: gen3-client/jwt/functions.go [hunk 7]
```diff
@@ -184,8 +180,54 @@ func (f *Functions) DoRequestWithSignedHeader(profile string, configFileType str
...
+	}
+
+	project_access, ok := data["project_access"].(map[string]interface{})
+	if !ok {
+		panic("Not possible to read user access privileges")
```
Comment (mfshao) [inline gen3-client/jwt/functions.go]: same as above

File: gen3-client/g3cmd/auth.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/spf13/cobra"
+	"github.com/uc-cdis/gen3-client/gen3-client/jwt"
+)
+
+func init() {
+
+	var authCmd = &cobra.Command{
+		Use:     "auth",
+		Short:   "Return data access priveleges from profile",
+		Long:    `Gets data access priveleges for specified profile.`,
```
Comment (mfshao) [inline gen3-client/g3cmd/auth.go]: typo, should be `privileges`

File: gen3-client/g3cmd/auth.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+			request := new(jwt.Request)
+			configure := new(jwt.Configure)
+			function := new(jwt.Functions)
+
+			function.Request = request
+			function.Config = configure
+
+			endPointPostfix := "/user/user" // Information about current user
+
+			host, project_access, err := function.CheckPrivileges(profile, "", endPointPostfix, "application/json", nil)
```
[the remark this answers is not in the document]
Comment (vzpgb) [reply gen3-client/g3cmd/auth.go]: Thanks! Changed in the commit [0878755](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/08787553bc29e981c27d6efb6eef8ba50f39d9c7).

File: gen3-client/g3cmd/auth.go (hunk 1 above)
Comment (vzpgb) [reply gen3-client/g3cmd/auth.go]: Thanks! Changed in the commit 0878755.

File: gen3-client/g3cmd/auth.go (hunk 2 above)
Comment (vzpgb) [reply gen3-client/g3cmd/auth.go]: Good idea! Implemented in commit [8cc9275] (https://github.com/uc-cdis/cdis-data-client/pull/49/commits/8cc92755aa94a9dbeb0c12e01c894a06ecf58271).

File: gen3-client/jwt/functions.go (hunk 3 above)
Comment (vzpgb) [reply gen3-client/jwt/functions.go]: Thanks! Fixed in the commit [33dfedf](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/33dfedfc97fbef84fe5d7a2465d075d60aa6d4dd)

File: gen3-client/jwt/functions.go (hunk 4 above)
Comment (vzpgb) [reply gen3-client/jwt/functions.go]: Thanks! Fixed in the commit [33dfedf](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/33dfedfc97fbef84fe5d7a2465d075d60aa6d4dd)

File: gen3-client/jwt/functions.go (hunk 6 above)
Comment (vzpgb) [reply gen3-client/jwt/functions.go]: Good to know, thanks! Fixed in the commit [33dfedf](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/33dfedfc97fbef84fe5d7a2465d075d60aa6d4dd)

File: gen3-client/jwt/functions.go (hunk 7 above)
Comment (vzpgb) [reply gen3-client/jwt/functions.go]: Also fixed in the commit [33dfedf](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/33dfedfc97fbef84fe5d7a2465d075d60aa6d4dd), thanks!

File: gen3-client/g3cmd/auth.go (hunk 8 above)
Comment (vzpgb) [reply gen3-client/g3cmd/auth.go]: Fixed in [0878755](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/08787553bc29e981c27d6efb6eef8ba50f39d9c7), thanks!

Comment (mfshao) [reply gen3-client/g3cmd/auth.go]: there is another typo in the line above this line

File: gen3-client/jwt/functions.go (hunk 5 above)
Comment (vzpgb) [reply gen3-client/jwt/functions.go]: Removed in the commit [835c73d](https://github.com/uc-cdis/cdis-data-client/pull/49/commits/835c73dad2f8d5e8d6754be9364117b0d864a35f). Thanks!

**Pull request merged**